Migrate block editor insert usage to preferences store

talldan opened the pull request:
Part of https://github.com/WordPress/gutenberg/issues/31965

## What?
The block editor package maintains some block insertion data to aid the 'Most used blocks' feature in the inserter. It persists this data.

This PR migrates the data to the preferences store.

## Why?
As outlined in https://github.com/WordPress/gutenberg/issues/31965, the goal is to move persisted data out of our various stores to a centralized preferences store.

## How?
<s>This required adding the preferences store as a dependency of the block editor package.</s>

The logic in the block editor's `preferences` reducer is now moved over to a new action `updateInsertUsage` that uses the preferences store. This new private action is dispatched in the `insertBlocks` and `replaceBlocks` actions.

The local storage data is migrated across to live under the preferences store from the old preferences system.

The preferences store's API usually uses a `scope` parameter that's usually set to 'core/edit-post' or 'core/edit-site', to indicate which editor the data belongs to. `insertUsage` stats are shared across editors, so I've used the 'core' scope.

Finally, this PR introduces a new private action in the preferences store, `markNextChangeAsExpensive`, which can be dispatched before updating a preference to delay the regularity with which the persistence layer makes HTTP requests to update user meta. It means you can insert loads of blocks quickly, but user meta will only update after a minute has passed. Insert usage still backs up immediately to a local storage persistence layer, so there's lots of resilience.

## Testing Instructions
1. Using `trunk` open the post editor
2. In the editor preferences modal, in the blocks section, enable the 'Show most used blocks' setting
3. Delete your WP local storage data and reload
4. Open the main inserter and click one of the blocks repeatedly (not paragraph, which is already first) in the Most Used section until it becomes the first block (note - possibly a bug that the order of blocks update while the inserter is open)
5. Checkout this branch and rebuild
6. Reload the post editor and open the main inserter, the order of the most used blocks should be retained (local storage data was migrated correctly)
7. Select a different block several times and make it the most used block.
8. Reload the post editor and open the main inserter, the order of the most used blocks should be retained (data is still persisted)

talldan committed:
Migrate block editor insert usage to preferences store

Update tests

Make updateInsertUsage a proper action that can be unit tested

Fix reusable block tests

Update test

Try fixing private actions with new store registration

Update tests

Add special handling for the insertUsage migration as it was performed later than others

Remove unused function

Add mark next change as expensive action to preferences store

Update debounce function to handle a longer debounce for expensive changes

Mark the insertUsage preference change as expensive

Make expensive calls use a trailing edge debounce

Fix duplicate keys in tests

Improve trailing edge test

Fix tests, and ensure options object is optional

Make updateInsertUsage a private API

Make markNextChangeAsExpensive a private API

Update docs

Update package-lock

Opt-in preferences as a core module using private apis

Do not unlock what is already unlocked

Remove time property from INSERT_BLOCKS and REPLACE_BLOCKS action objects

Rename `match` to `variation`

Rename file to create-async-debouncer

Add an extra param for defining the debounce of expensive requests

Add a default value of `false` for the `isExpensive` option

Make `__unstableGetInsertUsage` a private selector called `getInsertUsage`. Remove `__unstableGetInsertUsageForBlock`

Only run the migration when needed

Keep the `insertUsage` data at a maximum of 100 records

Fix linting issues

Update docs

talldan commented:
Looks like there are some failing tests - some more unit tests that need various mocks in place now that they touch the preferences store. I'll update those tomorrow.

talldan commented on packages/reusable-blocks/src/store/test/actions.js:
```diff
@@ -31,6 +32,7 @@ function createRegistryWithStores() {
 	registry.register( blockEditorStore );
 	registry.register( reusableBlocksStore );
 	registry.register( blocksStore );
+	registry.register( preferencesStore );
```
The issue here is that the tests fail because the block editor store uses the preferences store, but it isn't part of the test registry.

Here I'm adding it to the registry, but technically `preferences` isn't a dependency of the `reusable-blocks` package, so importing it doesn't seem right. I'm surprised the linter doesn't show an error.

I'm not sure of the correct way to resolve this issue. Maybe a mock preferences store would be better.

@adamziel I was wondering if you've encountered this situation before and have any suggestions?

adamziel replied:
@talldan my mental model here is that the `preferences` store actually is a dependency of a `reusable-blocks`, only not a direct one. It is a transitive dependency we get from the `blockEditorStore`. I don't think we have a canonical way of dealing with that when mocking test registries at the moment, but it would be great to have one. 

Thinking out loud, maybe the store object could have a `dependentStores` property with a list of any dependent stores? That way the tests could iterate over that tree recursively without introducing package.json-level dependencies. Even better if there was a helper function like `mockRegistryWithStores([ blockEditorStore, ... ])`

talldan commented:
There's a discussion in https://github.com/WordPress/gutenberg/pull/39672#discussion_r832967226 about whether `preferences` should be a dependency of `block-editor`. In the case of this PR, it's difficult to avoid that.

tellthemachines commented on packages/block-editor/package.json:
```diff
@@ -52,6 +52,7 @@
 		"@wordpress/keyboard-shortcuts": "file:../keyboard-shortcuts",
 		"@wordpress/keycodes": "file:../keycodes",
 		"@wordpress/notices": "file:../notices",
+		"@wordpress/preferences": "file:../preferences",
```
I'm thinking that having preferences as a dependency of the block editor means we'll have to be super careful not to make it WP-specific.

tellthemachines commented on packages/block-editor/src/store/defaults.native.js:
```diff
@@ -23,4 +20,4 @@ const SETTINGS_DEFAULTS = {
 	},
 };
 
-export { PREFERENCES_DEFAULTS, SETTINGS_DEFAULTS };
```
I can't work out if this was actually being used anywhere. It doesn't seem to be imported from any other native files 🤔

tellthemachines commented on packages/data/src/plugins/persistence/index.js:
```diff
@@ -427,24 +434,29 @@ persistencePlugin.__unstableMigrate = ( pluginOptions ) => {
 	);
+	migrateIndividualPreferenceToPreferencesStore(
+		persistence,
+		{ from: 'core/block-editor', scope: 'core' },
```
Couldn't scope be `'core/block-editor'`?

talldan replied:
I don't know that there's any advantage to grouping preferences per package in that way now.

The advantage of using 'core' is that if the feature is moved between packages, the scope still makes sense and a data migration isn't needed.

I think it'll be good to use 'core' for anything shared across editors in this way.

I should update the docs to make this usage clear. I'll work on a separate PR for that.

tellthemachines replied:
> The advantage of using 'core' is that if the feature is moved between packages, the scope still makes sense and a data migration isn't needed.

Not sure I understand. Would we be moving it from preferences to somewhere else?

> I think it'll be good to use 'core' for anything shared across editors in this way.

Do we really want the usage to be shared across editors though? It's unlikely that most used/most relevant blocks in the post editor will be the same as in the site editor.

talldan replied:
> Not sure I understand. Would we be moving it from preferences to somewhere else?

It's not so much the preferences, but the code that interacts with the preferences store could be moved. As an example, the code that tracks `insertUsage` has already been moved in the past from the `editor` package to `block-editor`. If it moved from `block-editor` to elsewhere again in the future it's better not to have to change the scope.

The scope is more related to how widely the data is used rather than where the feature is implemented.

> Do we really want the usage to be shared across editors though? It's unlikely that most used/most relevant blocks in the post editor will be the same as in the site editor.

It's already shared across editors. It's a good point, that could be changed, but not something I'd tackle in this PR.

tellthemachines replied:
> If it moved from block-editor to elsewhere again in the future it's better not to have to change the scope.

Got it, that makes sense.

tellthemachines commented on packages/block-editor/src/store/test/actions.js:
```diff
@@ -595,12 +642,206 @@ describe( 'actions', () => {
+									return { name: 'orange' };
+								if ( fruit === 'apple' )
+									return { name: 'apple' };
+							}
```
The added complexity to these tests is not great, but I don't really have any better suggestions for now 😅

tellthemachines reviewed (commented):
Tested and all works well, both migrating usage from trunk and persisting new usage ✅ 

> possibly a bug that the order of blocks update while the inserter is open

If it is a bug, it's a very fun one 😄 

Couple questions below, code looks good otherwise.

talldan commented:
Thanks for the review! I'll hold off on merging this and switch it to a draft until this conversation about adding preferences as a dependency of block-editor (https://github.com/WordPress/gutenberg/pull/39672#discussion_r832967226) is resolved.

talldan commented:
Status update on this one, this conversation is very relevant to this PR - https://github.com/WordPress/gutenberg/pull/39795#issuecomment-1111800556.

Now that preferences are saved via the REST API, merging this would mean that every block insert would result in an API request, which is probably undesirable.

I'm going to be AFK for a month so won't be able to solve this one quickly, but happy to look at it when I return.

jsnajdr commented on packages/block-editor/src/store/index.js:
```diff
@@ -32,14 +32,8 @@ export const storeConfig = {
-const registeredStore = registerStore( STORE_NAME, {
-	...storeConfig,
-	persist: [ 'preferences' ],
-} );
```
This is why I'm interested in finishing this PR 🙂 I'd like to remove all usages of the deprecated `registerStore` function and the `persist` plugin, and this is the last existing instance.

jsnajdr commented:
> Now that preferences are saved via the REST API, merging this would mean that every block insert would result in an API request, which is probably undesirable.

We could solve this by making the `debouncedApiFetch` more flexible. `persistenceLayer.set` would be called not only with the entire new preferences object, but also with the array of keys we're changing. The `requestDebounceMS` option, for the persistence layer constructor function, would be either a number or a function, from `keys[]` to `number`. When setting `core.insertUsage`, it would wait one minute before saving, for other keys it would wait the default 2.5s.

That's a rather elegant API, not a hack, is backward compatible, and solves the problem neatly.

One thing that is unfortunate is that the persistence setup (`__unstableCreatePersistenceLayer`, `setPersistenceLayer`) is done in a PHP inline script in Core, the Gutenberg plugin doesn't control it. Makes it harder to change.

talldan commented:
> One thing that is unfortunate is that the persistence setup (__unstableCreatePersistenceLayer, setPersistenceLayer) is done in a PHP inline script in Core, the Gutenberg plugin doesn't control it. Makes it harder to change.

Yeah, that does make it a little difficult. Preferences doesn't know what kind of persistence layer it has, and the persistence layer doesn't know what kind of preferences it has. It's decoupled in that way so that the preferences package can be consumed by packages that are not WordPress specific.

It's particularly tricky, because the code that knows about an individual preference like `insertUsage` won't know what kind of persistence is being used (it could be local storage only where this isn't a concern, or it could be local storage and a REST API like WordPress uses). I think the code would have to communicate with the persistence layer through the preferences API.

It could work by being very generic, where maybe the preferences API only knows about concepts like 'local' and 'remote', and the implementer of the persistence layer can interpret that as they want.

It could even be something very basic where a particular preference is marked as being expensive.

jsnajdr commented:
> It could even be something very basic where a particular preference is marked as being expensive.

Yes, this could work: mark the option declaratively and then the persistence layer can decide what to do about it. Like storing it in local storage in real time, but over network only in longer intervals.

talldan commented:
I think I have something working here.

- I've added a `markNextChangeAsExpensive` action to the preferences store. This gets called right before `insertUsage` is updated. 
- The persistence layer receives this as an `isExpensive` option in its `set` function.
- Expensive preferences result in a one minute trailing edge debounce, so they don't happen immediately, unlike inexpensive preferences.

I think the only thing that doesn't feel right is that when you make multiple 'expensive' preference changes, and then follow that up with an inexpensive preference change, there's still a one minute wait. Potentially the inexpensive change could cancel the expensive one. 🤔 

This new `markNextChangeAsExpensive` action isn't private right now, but I think I'll make it private. I went with this over a declarative approach, as it seems more minimal. I also didn't want the persistence layer to have to reason about individual preferences, as it's more of a black box. Though now I look at it I realise the preferences package could make this judgement, so perhaps I should still explore this option 😄 

I can look into splitting the preference API changes out into a separate PR for easier review if that helps.

talldan commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -1780,19 +1783,43 @@ export function canLockBlockType( state, nameOrType ) {
+ *                                 of block and an object value that contains
+ *                                 block insertion statistics.
+ */
+export const __unstableGetInsertUsage = createRegistrySelector(
```
I should make the __unstable selectors private instead (the PR originally predated the concept of private selectors)

noisysocks replied:
It's not a particularly complicated function. Instead of having it private / unstable could any callers that need it just call `select( preferencesStore ).get` directly?

Alternatively it can stay as a non-exported function but have it so that callers pass in `select` as an argument.

Same for `__unstableGetInsertUsageForBlock`.

talldan commented:
I've rebased and made a few extra changes:
- `updateInsertUsgae` is now a private action, as it's an internal implementation detail
- `markNextChangeAsExpensive` is also private

I don't think I'll have time to explore alternative options for a while, as I have some lengthy leave coming up, but now that the API is private I think it's ok to look at change it later down the line.

> I think the only thing that doesn't feel right is that when you make multiple 'expensive' preference changes, and then follow that up with an inexpensive preference change, there's still a one minute wait. Potentially the inexpensive change could cancel the expensive one. 

I think I was wrong on this, from testing again, it looks like a non-expensive action will cancel the expensive one, which is good.

noisysocks commented:
> insertUsage stats are shared across editors, so I've used the 'core' scope.

Thoughts on using `core/block-editor` here since the inserter belongs to `@wordpress/block-editor`?

noisysocks commented:
> I think I was wrong on this, from testing again, it looks like a non-expensive action will cancel the expensive one, which is good.

The data will be eventually persisted in either case, no?

talldan commented:
> Thoughts on using core/block-editor here since the inserter belongs to @wordpress/block-editor?

It means that if the preference ever moves packages, then you have to migrate it, which feels a bit unnecessary seeing as there would be no real difference in implementation.

I think the scopes are really different to the concept of packages, and more like a Venn diagram of which parts of WordPress the functionality encompasses.

talldan commented:
> The data will be eventually persisted in either case, no?

Yeah, it always sends through the latest full payload, so a later update will contain the data from any earlier cancelled update.

noisysocks commented:
Should we not persist `insertUsage` if _Show most used blocks_ is turned off? Because this setting is off by default I imagine the vast majority of users won't have it turned on in which case all of these POST requests will be happening unnecessarily. 

I'm also wondering if we need some way of limiting the amount of data that ends up in `insertUsage`. Right now `insertUsage` grows unboundedly. If you keep inserting new blocks then the object gets bigger and bigger. This is probably fine when it uses `localStorage` because _eventually_ the browser's storage will be cleared but now we will be storing this in the site's MySQL database and sending/receiving the object every time we make a request to `/me`.

Ideally we'd calculate the most recently used blocks by looking at e.g. only the last 30 days of usage, or something like that. That would also mitigate privacy concerns around storing usage data indefinitely.

noisysocks commented on packages/block-editor/src/store/actions.js:
```diff
@@ -1214,7 +1217,6 @@ export function replaceInnerBlocks(
 		blocks,
 		updateSelection,
 		initialPosition: updateSelection ? initialPosition : null,
-		time: Date.now(),
```
Can we also remove `time` from the `INSERT_BLOCKS` action object?

talldan replied:
Yeah, that's curious. Both `REPLACE_BLOCKS` and `INSERT_BLOCKS`. I wonder why `REPLACE_INNER_BLOCKS` has it anyway, as it doesn't update insert usage.

I'll push a commit that removes it entirely.

noisysocks commented on packages/block-editor/src/store/private-actions.js:
```diff
@@ -301,3 +303,52 @@ export function toggleRemovalPromptSupport( status = true ) {
+		const updatedInsertUsage = blocks.reduce( ( previousState, block ) => {
+			const { attributes, name: blockName } = block;
+			let id = blockName;
+			const match = registry
```
nit: renaming `match` to `variation` would make this easier to read.

talldan commented:
> Should we not persist insertUsage if Show most used blocks is turned off?

I think you still get the most used blocks in the quick and slash inserters even with that turned off, I must have forgotten about that when I wrote the testing instructions.

> I'm also wondering if we need some way of limiting the amount of data that ends up in insertUsage.

This is a good idea! What do you think is a good limit? Maybe 100. 

edit: Oh, you said 30 days worth, yep that would work.

noisysocks commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -2089,7 +2119,7 @@ export const getInserterItems = createSelector(
 		state.blocks.byClientId,
 		state.blocks.order,
-		state.preferences.insertUsage,
+		__unstableGetInsertUsage(),
```
Double checking: are the objects returned by `get` stable? Referential equality is used by `rememo` to determine if cache should be invalidated.

talldan replied:
The values are from the preferences store, and those values only change when the `insertUsage` preference is updated, so should be no issues.

noisysocks commented on packages/preferences-persistence/src/create/debounce-async.js:
```diff
@@ -9,23 +9,21 @@
+ *                    options parameter.
  */
-export default function debounceAsync( func, delayMS ) {
+export default function createAsyncDebouncer() {
```
nit: now the filename doesn't match the default export.

talldan replied:
Renamed 👍

noisysocks commented on packages/preferences-persistence/src/create/index.js:
```diff
@@ -68,7 +68,7 @@ export default function create( {
 	}
 
-	function set( newData ) {
+	function set( newData, { isExpensive } = {} ) {
```
nit

```suggestion
	function set( newData, { isExpensive = false } = {} ) {
```

noisysocks commented on packages/preferences-persistence/src/create/index.js:
```diff
@@ -83,26 +83,28 @@ export default function create( {
+					},
+				} ).catch( () => {} ),
+			{
+				delayMS: isExpensive ? 60000 : requestDebounceMS,
```
Should `60000` be configurable via a param passed to `create` same as `requestDebounceMS` is?

noisysocks commented on packages/preferences-persistence/src/index.js:
```diff
@@ -38,9 +42,27 @@ export function __unstableCreatePersistenceLayer( serverData, userId ) {
+
+	// The insertUsage preference was migrated later than others, the following code
+	// performs a separate migration just for this data.
+	if ( preloadedData && ! preloadedData?.core?.insertUsage ) {
```
Will this run again and again if `insertUsage` is falsey because the user never has inserted a block?

talldan replied:
Hmmmmm. Yes. I think you're right.

I added an inner `if` statement to check for data to migrate before doing anything more expensive.

noisysocks commented:
> I think you still get the most used blocks in the quick and slash inserters even with that turned off, I must have forgotten about that when I wrote the testing instructions.

Ah cool.

> > I'm also wondering if we need some way of limiting the amount of data that ends up in insertUsage.
> 
> This is a good idea! What do you think is a good limit? Maybe 100.
> 
> edit: Oh, you said 30 days worth, yep that would work.

My saying 30 days came from nowhere 😅 I don't think it really matters how we do it just so long as the object doesn't grow unboundedly as users install and insert new blocks. Sorting the key/value pairs and then calling `.slice( 100 )` also works.

talldan commented:
> It's not a particularly complicated function. Instead of having it private / unstable could any callers that need it just call select( preferencesStore ).get directly?

> Alternatively it can stay as a non-exported function but have it so that callers pass in select as an argument.

> Same for __unstableGetInsertUsageForBlock.

Yeah, I don't like it, the reason it's there is the usage in the `getDependants` function of the memoized `getInserterItems` selector:
```js
export const getInserterItems = createSelector(
	( state, rootClientId = null ) => {
	        // ...
	},
	( state ) => [
	        // ...,
		__unstableGetInsertUsage(),
	]
);
```

I don't think there's a way to call `select` properly, and it's a registry selector so it has to be exported to be bound.

My understanding is that this PR will unlock the ability to use private registry selectors in the block editor store, so I've switched it to being a private registry selector called `getInsertUsage` and deleted `__unstableGetInsertUsageForBlock`.

I don't know if there's another way around this, but this provides some nice symmetry with `updateInsertUsage` being a private action, and the tests are also better now.

talldan commented:
> My saying 30 days came from nowhere 😅 I don't think it really matters how we do it just so long as the object doesn't grow unboundedly as users install and insert new blocks. Sorting the key/value pairs and then calling .slice( 100 ) also works.

I tried limiting it to the 100 most recently inserted blocks. I did it this way so that if a person goes away for three months, and then comes back they still have some `insertUsage` data.

I think 100 should be enough to not have an adverse effect on the frecency. I did also consider only keeping the 100 most frecent blocks, but this causes an issue in that new blocks can never get into an established top 100 (since they need some existing `insertUsage` data for their count to increase), so I think using keeping 100 based on `time` is the right approach.

Since the slicing happens in `updateInsertUsage`, we should check the performance results.

I think I've addressed all the review feedback now.

jsnajdr commented on packages/block-editor/src/store/actions.js:
```diff
@@ -603,12 +604,12 @@ export const insertBlocks =
 			}
 		}
 		if ( allowedBlocks.length ) {
+			dispatch( updateInsertUsage( blocks ) );
```
This should be `updateInsertUsage( allowedBlocks )`. Only the allowed ones are actually inserted.

Also, for consistency with `REPLACE_BLOCKS`, let's update insert usage only _after_ the actual insert is done.

jsnajdr commented on packages/block-editor/src/store/actions.js:
```diff
@@ -398,11 +399,11 @@ export const replaceBlocks =
 			initialPosition,
 			meta,
 		} );
+		dispatch( updateInsertUsage( blocks ) );
```
You should be able to call `dispatch.updateInsertUsage( blocks )`. Since #51051, unlocked private actions are exposed to thunks, on the `dispatch` object. See also #52214.

jsnajdr commented on packages/block-editor/src/store/private-actions.js:
```diff
@@ -289,3 +291,71 @@ export function toggleRemovalPromptSupport( status = true ) {
+			entries.sort( ( entryLeft, entryRight ) => {
+				const [ , { time: timeLeft } ] = entryLeft;
+				const [ , { time: timeRight } ] = entryRight;
+				return timeRight - timeLeft;
```
I'd rather write the callback as simple `( a, b ) => b[ 1 ].time - a[ 1 ].time`.

jsnajdr commented on packages/block-editor/src/store/selectors.js:
```diff
@@ -2008,7 +1996,8 @@ export const getInserterItems = createSelector(
-			const { time, count = 0 } = getInsertUsage( state, id ) || {};
+			const insertUsage = getInsertUsage();
+			const { time, count = 0 } = insertUsage?.[ id ] ?? {};
 			const frecency = calculateFrecency( time, count );
```
The `getInsertUsage` and `calculateFrecency` calls could be one day combined into one `getBlockFrecency( id )` selector. They are always used together, there's no other use of the insert usage data than calculating frecency for a particular block.

jsnajdr commented on packages/preferences-persistence/src/create/create-async-debouncer.js:
```diff
@@ -9,23 +9,21 @@
 		// This is a leading edge debounce. If there's no promise or timeout
 		// in progress, call the debounced function immediately.
-		if ( ! activePromise && ! timeoutId ) {
+		if ( ! isTrailing && ! activePromise && ! timeoutId ) {
```
I think this debounce logic has been a little bit flawed from the beginning, and the long `expensiveRequestDebounceMS` makes it worse.

If I call the debounced function repeatedly, the timeout keeps getting cancelled and rescheduled, and as a result the save can be never performed. Or only after a very long time, much longer than the intended timeout.

To avoid this "starvation" problem, we should do throttling rather than debouncing. Once the 60sec timeout is active, keep it running, but on further calls only update the _function_ to be called when it finally expires.

The Lodash [`debounce`](https://lodash.com/docs/4.17.15#debounce) function has a `maxWait` option that achieves something similar.

If we fix this behavior, then maybe we can remove also the invocation on the leading edge? I don't know why it was originally added -- why can't we always wait 2.5s? Maybe it's a workaround for this starvation problem?

tyxla replied:
Just noting that we already have both `throttle` and `debounce` utilities as part of the `@wordpress/compose` package, and the built-in `debounce()` also supports a `maxWait` option. Just in case that could be something you'd like to utilize instead of a custom debounce logic.

talldan replied:
> If I call the debounced function repeatedly, the timeout keeps getting cancelled and rescheduled, and as a result the save can be never performed. Or only after a very long time, much longer than the intended timeout.
>To avoid this "starvation" problem, we should do throttling rather than debouncing. Once the 60sec timeout is active, keep it running, but on further calls only update the function to be called when it finally expires.

I'll look into it. It's not really much of an issue with the shorter delay, but perhaps becomes more apparent with the longer delay. 🤔 

The main aim is to avoid repeated REST requests. WordPress servers are very bad at handling multiple requests at the same time, but perhaps it's ok to throttle.

> If we fix this behavior, then maybe we can remove also the invocation on the leading edge? I don't know why it was originally added -- why can't we always wait 2.5s? Maybe it's a workaround for this starvation problem?

The most common use case is infrequent preferences calls, so triggering the request right away is usually fine. It also reduces the chances that the delay/request might be in flight while the user unloads the page, causing the request to never go through.

> Just noting that we already have both throttle and debounce utilities as part of the @wordpress/compose package, and the built-in debounce() also supports a maxWait option. Just in case that could be something you'd like to utilize instead of a custom debounce logic.

@tyxla It needs to support async functions, which most libraries don't. The code here ensures requests are never made in parallel—it delays from after a promise completion rather than from function invocation. It might be there's another way of doing this that allows leverage of worpdress/compose, any suggestions appreciated.

talldan replied:
I've updated it to work as @jsnajdr describes so that we can try it out.

I think it works well. I've kept the debounce for the non-expensive call a short 1 second, and the expensive one 5 seconds, and there's a 10 second max wait.

The values feel a bit arbitrary, but I figure it should be ok to prevent spamming of HTTP requests if the user is duplicating lots of blocks.

The function is now also very simple, which is a bonus.

I think all the other feedback has also been addressed. I still need to give this a thorough test as so long has passed since the last time I looked at this PR.

jsnajdr commented on packages/block-editor/src/store/private-actions.js:
```diff
@@ -289,3 +291,71 @@ export function toggleRemovalPromptSupport( status = true ) {
+			updatedInsertUsage = Object.fromEntries( entriesToKeep );
+		}
+
+		registry.dispatch( preferencesStore );
```
Did you forget to call `markNextChangeAsExpensive` here? You'll need to also `unlock` the `dispatch` object first: `unlock( registry.dispatch( prefStore ) ).mark()`.

Alternatively, we could introduce an `options` fourth argument for `set`. Or have a private `.expensiveSet` action.

jsnajdr reviewed (commented):
I like this a lot and am looking forward to see it merged! I pointed out a few bugs and nits, but conceptually this is perfectly OK IMO.

talldan commented:
I don't have any time to work on this in the forseeable future, so would be happy for someone to volunteer to take this over.

talldan committed:
Fix block editor actions tests

talldan committed:
Fix frecency test

talldan committed:
Minimize properties in test assertions

talldan committed:
Ensure markNextChangeAsExpensive is called and update tests

talldan committed:
Fix only track allowedBlocks in insertBlocks action

talldan committed:
Simplify sort callback

talldan committed:
Update debouncing function - use trailing edge debounce only and implement a maxWait

talldan committed:
Fix persistence layer tests

talldan committed:
Update docs

talldan committed:
Fix resetting of preferences in e2e test

talldan committed:
Hide the welcome guide in e2e test

talldan commented:
The changes to debouncing caused some interesting issues in e2e tests.

A couple of e2e tests use the preferences API to set up a test state (`setFixedToolbar( true )`), and then call it again at the end of the test case to tear down (`setFixedToolbar( false )`). The switch to only using a trailing edge debounce results in the second call never triggering an HTTP request, and on the starting the next test case the `true` state would be restored from user meta. 

Instead I've switched to `resetPreferences` which bypasses the preferences API (it updates user meta directly and playwright will wait for it to complete before moving on, while the other way is a 'fire and forget').

It also looks like preferences state is only ever reset at the start of running an entire test suite, not before individual files or test cases, so there can be lingering preferences state. Surprisingly it doesn't cause that many issues. 🤔

talldan commented on test/e2e/specs/site-editor/activate-theme.spec.js:
```diff
@@ -36,6 +36,12 @@ test.describe( 'Activate theme', () => {
+		// Usually `visitSiteEditor` is used to test the site editor, and it
+		// hides the welcome guide. Since this test uses a different flow,
+		// manually update the preferences here.
+		await editor.setPreferences( 'core/edit-site', {
+			welcomeGuide: false,
+		} );
```
This test actually fails in `trunk` if you run it locally. It only passes in CI in trunk because other tests that run before it set `welcomeGuide` to `false` 😬 

I'm not sure why it started failing on CI in this PR.

Mamaduka commented:
I think the mentioned end-to-end tests are flaky even on trunk. States like "Fixed Toolbar" can persist between tests, causing follow-up tests to fail. I'll try to make them more resilient separately, which should help partially unblock this.

It would be nice to resume work on this PR, the `block-editor` only major package, still using the "legacy" persistence system.

talldan commented:
> It would be nice to resume work on this PR, the block-editor only major package, still using the "legacy" persistence system.

Yeah, I'd love to get this across the line. Maybe once 6.9 stuff is out of the way I can picked it up again.
